src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs

```diff
@@ -148,28 +168,59 @@ internal TryCatchFinallyHandler(int tryStart, int tryEnd, int gotoEndLabelIndex,
             }
         }
 
-        /// <summary>
-        /// Goto the index of the first instruction of the suitable catch block
-        /// </summary>
-        internal int GotoHandler(InterpretedFrame frame, object exception, out ExceptionHandler handler)
+        internal bool HasHandler(InterpretedFrame frame, Exception exception, out ExceptionHandler handler)
         {
-            Debug.Assert(_handlers != null, "we should have at least one handler if the method gets called");
+            frame.SaveTraceToException(exception);
+
+            if (IsCatchBlockExist)
+            {
+                for(int i = 0; i != _handlers.Length; ++i)
+                {
+                    ExceptionHandler candidate = _handlers[i];
+                    if (candidate.Matches(exception.GetType()) && (candidate.Filter == null || FilterPasses(frame, exception, candidate.Filter)))
+                    {
+                        handler = candidate;
+                        return true;
+                    }
+                }
+            }
+
             handler = null;
-            for (int i = 0; i < _handlers.Length; i++)
+            return false;
+        }
+
+        internal bool FilterPasses(InterpretedFrame frame, Exception exception, ExceptionFilter filter)
+        {
+            Debug.Assert(filter != null);
+            Interpreter interpreter = frame.Interpreter;
+            Instruction[] instructions = interpreter.Instructions.Instructions;
+            int stackIndex = frame.StackIndex;
+            try
             {
-                if (_handlers[i].Matches(exception.GetType()))
+                int index = interpreter._labels[filter.LabelIndex].Index;
+                frame.Push(exception);
+                while (index >= filter.FilterStartIndex && index < filter.FilterEndIndex)
                 {
-                    handler = _handlers[i];
-                    break;
+                    index += instructions[index].Run(frame);
                 }
+
+                object result = frame.Pop();
+                Debug.Assert(result is bool);
+                Debug.Assert(frame.StackIndex == stackIndex);
+                return (bool)result;
+            }
+            catch(Exception)
+            {
```

Comment:
It makes little odds, but that would be a bit clearer as to our ignoring the exception (at first I'd intended to have DEBUG builds rethrow if it was an assertion failure, but that was getting messy for little gain, so I'd had `Exception ex` in there for that). Will change to empty catch.